pkg/store/proxy.go

```diff
@@ -37,6 +37,16 @@ type ctxKey int
 // StoreMatcherKey is the context key for the store's allow list.
 const StoreMatcherKey = ctxKey(0)
 
+// StoreType determines if the store for a given client is a local store
+// (i.e. using server-as-client with in-process stream) or a store to which
+// the client connects remotely over gRPC.
+type StoreType string
```

Comment:
Why we need to differntiate?

Reply:
In line with discussion in https://github.com/thanos-io/thanos/issues/2864, I also felt the current interface does fit that good with the local client scenario. Currently, when calling `Addr()` we should get a valid `host:port` address. 

In case of local client, however, there is realistically no address - we're not calling any address, we're not calling `localhost`, we're using the server-as-client implementation, meaning we're not really calling any address but directly streaming the series. 

So we could hide this away and make up an address, but that does not seem right to me since there is not really an address attached to the local client. Maybe other alternative would be to not return any address or return some other string (e.g. `<local_client>`) to indicate that the address is not applicable. Then we could avoid having this type but it seems still somewhat less clear than having this explicitly stated.

I'm open to this or other alternative solutions / opinions though.

Reply:
Yea, let me check next week, but I see your point. Another way would be to create `Addr() (addr string, noAddress bool)` which is simple and self-explanatory.

Reply:
That could also work :+1: I'm open to whatever is best understandable for others :blush:

Reply:
@fpetkovski pretty much, filtering and I guess debugging, tracing tags etc.

I see your points, after more thinking I'd agree that it might be overkill to have that type since it's basically either one or the other, so something along what @bwplotka suggested should be clear enough, I'll remake this.

Reply:
Sounds like we need to address it still.

Reply:
Addressed :+1: